### PR TITLE
feat: access handler name add properties to req route options

### LINKED ISSUE #4439: Access handler name
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the feature has not already been requested

### 🚀 Feature Proposal

Open Telemetry accesses the (now deprecated via #4216) `request.context` property so it can try to get the name of the handler function. See https://github.com/open-telemetry/opentelemetry-js-contrib/blob/4cafe6becc19cfb76c53393cb03a5c0574a68b0d/plugins/node/opentelemetry-instrumentation-fastify/src/instrumentation.ts#L264-L268

With `request.context` deprecated, we now get a big deprecation warning printed every the app gets its first request. Would it be possible to expose this information in a meaningful way?

There's an issue in their repo as well: https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1275

### Motivation

It's useful to get the name of a handler function as a span name in traces.

### Example

Would replace `request.context.handler.name`

### PR DESCRIPTION
**Pull request by cesarvspr**
Closes #4439

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Review by Eomm — COMMENTED** [review]
The `request.routeOptions should be immutable` test is failing, it could be due to the new warning

File: lib/request.js [hunk 1]
```diff
@@ -171,6 +172,7 @@ Object.defineProperties(Request.prototype, {
       const routeLimit = context._parserOptions.limit
       const serverLimit = context.server.initialConfig.bodyLimit
       const version = context.server.hasConstraintStrategy('version') ? this.raw.headers['accept-version'] : undefined
+      const { handler, config } = context
```
Comment (climba03003) [inline lib/request.js:175]: You are taking the reference here and `config` is a object.
When you deep frozen the object, altering `config` in else where will be break.

Comment (cesarvspr) [reply lib/request.js]: pal

File: lib/request.js [hunk 2]
```diff
@@ -179,7 +181,10 @@ Object.defineProperties(Request.prototype, {
         logLevel: context.logLevel,
         exposeHeadRoute: context.exposeHeadRoute,
         prefixTrailingSlash: context.prefixTrailingSlash,
-        version
+        handler,
+        version,
+        config,
+        schema
```
[the remark this answers is not in the document]
Comment (cesarvspr) [reply lib/request.js]: ptal

File: test/router-options.test.js [hunk 3]
```diff
@@ -104,9 +104,12 @@ test('Should expose router options via getters on request and reply', t => {
   }, (req, reply) => {
     t.equal(reply.context.config.url, '/test/:id')
     t.equal(reply.context.config.method, 'GET')
-    t.equal(req.routeConfig.url, '/test/:id')
-    t.equal(req.routeConfig.method, 'GET')
-    t.same(req.routeSchema, expectedSchema)
+    t.equal(req.routeConfig.url, '/test/:id') // TODO: deprecate on @v5
+    t.equal(req.routeConfig.method, 'GET') // TODO: deprecate on @v5
+    t.same(req.routeSchema, expectedSchema) // TODO: deprecate on @v5
```
Comment (Eomm) [inline test/router-options.test.js:109]: We should move these assertions to a test case that does the same check as https://github.com/fastify/fastify/blob/main/test/listen.deprecated.test.js

File: lib/request.js [hunk 4]
```diff
@@ -191,11 +203,13 @@ Object.defineProperties(Request.prototype, {
   },
   routeConfig: {
     get () {
+      warning.emit('FSTDEP015')
       return this[kRouteContext][kPublicRouteContext].config
     }
   },
   routeSchema: {
     get () {
+      warning.emit('FSTDEP014')
```
Comment (Eomm) [inline lib/request.js:212]: We need to test is as https://github.com/fastify/fastify/blob/main/test/listen.deprecated.test.js

File: lib/request.js [hunk 5]
```diff
@@ -179,8 +180,19 @@ Object.defineProperties(Request.prototype, {
         logLevel: context.logLevel,
         exposeHeadRoute: context.exposeHeadRoute,
         prefixTrailingSlash: context.prefixTrailingSlash,
+        handler,
         version
       }
+
+      Object.defineProperties(options, {
+        config: {
+          get: () => config
```
Comment (Eomm) [inline lib/request.js:189]:
```suggestion
          get: () => context.config
```

File: lib/request.js [hunk 6]
```diff
@@ -179,8 +180,19 @@ Object.defineProperties(Request.prototype, {
...
+        },
+        schema: {
+          get: () => schema
```
Comment (Eomm) [inline lib/request.js:192]:
```suggestion
          get: () => context.schema
```

File: lib/request.js [hunk 7]
```diff
@@ -171,6 +171,7 @@ Object.defineProperties(Request.prototype, {
       const routeLimit = context._parserOptions.limit
       const serverLimit = context.server.initialConfig.bodyLimit
       const version = context.server.hasConstraintStrategy('version') ? this.raw.headers['accept-version'] : undefined
+      const { handler, config, schema } = context
```
Comment (Eomm) [inline lib/request.js:174]:
```suggestion
      const { handler } = context
```

File: test/router-options.test.js [hunk 8]
```diff
@@ -104,9 +104,12 @@ test('Should expose router options via getters on request and reply', t => {
...
+    t.same(req.routeOptions.schema, expectedSchema)
+    t.equal(req.routeOptions.config.url, '/test/:id')
+    t.equal(req.routeOptions.config.method, 'GET')
```
Comment (Eomm) [inline test/router-options.test.js:112]: We need to add the check for the `routeOptions.handler` here

File: lib/request.js (hunk 4 above)
Comment (cesarvspr) [reply lib/request.js:214]: created a file ptal

File: lib/request.js (hunk 5 above)
Comment (cesarvspr) [reply lib/request.js]: updated tks

File: lib/request.js (hunk 6 above)
Comment (cesarvspr) [reply lib/request.js]: updated tks

File: lib/request.js (hunk 7 above)
Comment (cesarvspr) [reply lib/request.js]: updated tks

File: test/router-options.test.js (hunk 8 above)
Comment (cesarvspr) [reply test/router-options.test.js:110]: updated tks

File: test/router-options.test.js (hunk 3 above)
Comment (cesarvspr) [reply test/router-options.test.js]: created a file ptal

File: test/request.deprecated.test.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
+'use strict'
+
+// Tests for some deprecated `request.*` keys. This file should be
+// removed when the deprecation is complete.
+
```
Comment (Eomm) [inline test/request.deprecated.test.js:5]: The scope of isolating the deprecated getter into a test is to suppress the warning when you run `npm test` locally.

So we need to fix this one too:

> test/route.test.js 2> (node:72195) [FSTDEP012] FastifyDeprecation: Request#context
property access is deprecated. Please use "Request#routeConfig" or
"Request#routeSchema" instead for accessing Route settings. The "Request#context"
will be removed in `fastify@5`.

This suggestion fixes the warning from this file

```suggestion

process.removeAllListeners('warning')

```

**Review by Eomm — COMMENTED** [review]
We need to update the documentation as well

**Comment by rohitcoder** [comment]
@Eomm I'm also getting this same error, and i wasn't able to find much info, how i can get this fixed.
Screenshot 2023-01-01 at 5 11 01 PM

**Comment by climba03003** [comment]
Please open an issue for you question or ask in `fastify/help`.

**Comment by cesarvspr** [comment]
> We need to update the documentation as well

I agree, are docs expected to be addressed in the same PR? @Eomm

**Comment by Eomm** [comment]
Yes, we keep all the things aligned in this way

**Comment by metcoder95** [comment]
Hi @cesarvspr, are there any updates on this? 🙂

**Review by mcollina — APPROVED** [review]
lgtm

**Review by mcollina — APPROVED** [review]
lgtm

**Review by mcollina — APPROVED** [review]
lgtm

**Review by rahmanvivasoft — APPROVED** [review]

**Comment by SimenB** [comment]
Ping 🙂

**Comment by Eomm** [comment]
Two comments to be addressed:

- https://github.com/fastify/fastify/pull/4470#issuecomment-1379758550
- https://github.com/fastify/fastify/pull/4470#discussion_r1058433703

then we can land this

**Comment by Songkeys** [comment]
What can we do to move this PR forward?

Currently my logging is full of this:

image

**Comment by Uzlopak** [comment]
Is this even related to this PR?

Cant you patch your code accordingly?

**Comment by SimenB** [comment]
>What can we do to move this PR forward?

From my understanding, only a doc update is missing. So you can probably suggest changes here or PR the fork.

**Comment by Songkeys** [comment]
> Is this even related to this PR?
>
> Cant you patch your code accordingly?

I think this is related to another issue for the opentelemetry which is blocked by this PR? https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1275

**Comment by Uzlopak** [comment]
I think @SimenB wanted to actually say, that this PR would also add more warnings to the opentelemetry issue.

**Comment by Uzlopak** [comment]
To fix the issue in opentelemetry, you would need to patch 

https://github.com/open-telemetry/opentelemetry-js-contrib/blob/154b30bc98ff0b848348faa18197eb7f11dcbdc4/plugins/node/opentelemetry-instrumentation-fastify/src/instrumentation.ts#L263

There I think you would need to change it to:
```js
      const handlerName = (((request as any).routeConfig || {})?.name || '').substr(6);
```

**Comment by Songkeys** [comment]
Just took another look at this PR- yes you're right. Sorry about this. Guess there's some work needed to do on the opentelemetry side then.

File: lib/request.js [hunk 10]
```diff
@@ -189,23 +191,37 @@ Object.defineProperties(Request.prototype, {
         logLevel: context.logLevel,
         exposeHeadRoute: context.exposeHeadRoute,
         prefixTrailingSlash: context.prefixTrailingSlash,
+        handler,
```
Comment (Uzlopak) [inline lib/request.js:194]:
```suggestion
        handler: context.handler,
```

File: lib/request.js [hunk 11]
```diff
@@ -181,6 +182,7 @@ Object.defineProperties(Request.prototype, {
       const routeLimit = context._parserOptions.limit
       const serverLimit = context.server.initialConfig.bodyLimit
       const version = context.server.hasConstraintStrategy('version') ? this.raw.headers['accept-version'] : undefined
+      const { handler } = context
```
Comment (Uzlopak) [inline lib/request.js:185]:
```suggestion
```

File: docs/Reference/Request.md [hunk 12]
```diff
@@ -47,12 +47,14 @@ Request is a core Fastify object containing the following fields:
...
 - `routeOptions` - The route [`option`](./Routes.md#routes-options) object
   - `bodyLimit` - either server limit or route limit
+  - `config` - the [`config`](./Routes.md#routes-config) object for this route
   - `method` - the http method for the route
   - `url` - the path of the URL to match this route
   - `handler` - the handler for this route
   - `attachValidation` - attach `validationError` to request 
     (if there is a schema defined)
   - `logLevel` - log level defined for this route
+  - `schema` - the scheme definition for this route
```
Comment (Eomm) [inline docs/Reference/Request.md:57]:
```suggestion
  - `schema` - the JSON schemas definition for this route
```

**Review by Eomm — APPROVED** [review]
We will need to check these warnings across the repos

Thanks GH code search

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**